RP1 clocks

6by9 opened the pull request:
For discussion over critical clocks, and those that shouldn't be disabled.

6by9 committed:
rp1: clk: Rationalise the use of the CLK_IS_CRITICAL flag

The clock setup had been copied from clk-bcm2835 which had to cope
with the firmware having configured clocks, so there were flags
of CLK_IS_CRITICAL and CLK_IGNORE_UNUSED dotted around.

That isn't the situation with RP1 where only the main PLLs, CLK_SYS,
and CLK_SLOW_SYS are critical, so update the configuration to match.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
rp1: clk: Remove CLK_IGNORE_UNUSED flags

There should be no issue in disabling the RP1 clocks as long as
the kernel knows about all consumers.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dt: arm64: Fixup RP1 ethernet DT configuration

Configure RP1's ethernet block to do the correct thing.
clk_eth is intended to be fixed at 125MHz, so use a new compatible,
and use assigned-clocks to configure the clock appropriately.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

P33M commented:
What happens if you force the ethernet link to 100Base-TX? The kernel shouldn't try to reconfigure clk_eth.

6by9 commented:
> What happens if you force the ethernet link to 100Base-TX? The kernel shouldn't try to reconfigure clk_eth.

Just following through on that one. Looks like it fails, so needs addressing.
 
> Edit: hclk and pclk are directly connected to clk_sys, but are still orphaned in that list above. Do you just declare those two as having clk_sys as the source?

I've not changed that. They're base clocks in https://github.com/raspberrypi/linux/blob/rpi-6.6.y/arch/arm64/boot/dts/broadcom/rp1.dtsi#L1196-L1207, hence being orphaned.
If they're actually off clk_sys, then the driver needs updating to add them.

P33M commented:
This is scary: `    clk_slow_sys                     0       0        0        50000000    0          0     50000      Y      deviceless                      no_connection_id         `
Which exists on downstream kernels too. Sure enough, its control register is zeroed. This should never be turned off.

6by9 commented:
> This is scary: ` clk_slow_sys 0 0 0 50000000 0 0 50000 Y deviceless no_connection_id ` Which exists on downstream kernels too. Sure enough, its control register is zeroed. This should never be turned off.

This is running on a downstream 6.6.60 kernel. It'll get rejigged for mainline afterwards.

Easy to add clk_slow_sys as a critical clock, but what does it feed?!

P33M commented:
Lots of important stuff that Linux doesn't know about. E.g. the downstream parts of the pad control registers (!) - which begs the question, how is pad control working at all?

P33M commented:
Ah. Someone's thought of that. Neither clk_sys nor clk_slow_sys have enable bits implemented in their control registers.

6by9 commented:
I'll make clk_sys and clk_slow_sys critical to make it more obvious what's going on.

I'm just observing that RP1_ETH_CLK does not have an assigned rate in rp1.dtsi. Presumably the 125MHz I'm seeing is therefore coming from the firmware config of it, but really ought to be configured by the kernel.

The simplest approach for ethernet looks to be looks to have the MACB_CAPS_CLK_HW_CHG flag set so `macb_set_tx_clk` aborts immediately, as used on `sama7g5_gem_config`. The description for daafa1d33cc9286d4b17ad87a18df55687fd0ab6 looks fairly close to our situation.
That does mean a custom compatible string though, so a bit of a faff. C'est la vie.

6by9 committed:
dt-bindings: net: cdns,macb: Add compatible for Raspberry Pi RP1

The Raspberry Pi RP1 chip has the Cadence GEM ethernet
controller, so add a compatible string for it.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
net: macb: Add support for Raspberry Pi RP1 ethernet controller

The RP1 chip has the Cadence GEM block, but wants the tx_clock
to always run at 125MHz, in the same way as sama7g5.
Add the relevant configuration.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Updated again.

New compatible on the ethernet MAC to set the MACB_CAPS_CLK_HW_CHG flag, so we can now change speed happily.

Any further comments?

6by9 commented:
Oh, we've still to resolve the question over hclk and pclk. Time to read the docs for them.

6by9 committed:
dt: rp1: Use clk_sys for ethernet hclk and pclk

hclk and pclk of the MAC are connected to clk_sys, so define
them as being connected accordingly, rather than having fake
fixed clocks for them.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
As discussed with Jonathan, hclk and pclk are connected to clk_sys, so redefined in DT to match that.

6by9 commented:
RP1 DMA (used by I2S) is doing funny things.

With "rp1: clk: Remove CLK_IGNORE_UNUSED flags", when `dma_request_chan` from `snd_dmaengine_pcm_register` results in a call to `dma_chan_alloc_chan_resources` and `axi_chan_is_hw_enable`, the return value of register `DMAC_CHEN` is 0xffffffff, when previously it returned 0x0. 

rp1_dma is defined in DT as using CLK_SYS and sdhci_core clocks, but CLK_SYS is already marked as critical, and sdhci_core is a fixed clock.
So is that correct, or is it using some other clock which we're now disabling as the kernel believes it is idle? (I'll got for a process of elimination to try and find which clock is required).

6by9 commented:
So the answer appears to be `RP1_PLL_SYS_PRI_PH` is the clock that is required for DMA to be able to read registers
Point DT for rp1_dma at that instead of RP1_SYS_CLK, and I2S works again.

The [binding for dw-axi-dmac](https://github.com/raspberrypi/linux/blob/rpi-6.6.y/Documentation/devicetree/bindings/dma/snps%2Cdw-axi-dmac.yaml#L44-L52) says core-clk should be the Bus Clock (currently pointing at `sdhci_core`), and cfgr-clk should be the Module Clock (was `RP1_CLK_SYS`). 
@P33M @naushir Is this correct?

P33M commented:
The DMAC has its own core clock divider, and I think "core" in the clk-summary list should be it

P33M commented:
The binding description is a bit mangled. There are no APB registers - at least not in our version of the hardware. There's just a flat set of configuration registers, clocked from clk_sys ("Bus clock" isn't a core clock). The internal state machines are/can be clocked from a different source - in our implementation, clk_dma ("Module clock" isn't a configuration clock).

6by9 committed:
clk: rp1: Add RP1_CLK_DMA.

The DMA block has a clock, but wasn't defined in the driver. This
resulted in the parent being disabled as unused, and then DMA
stopped working.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dt: rp1: Link RP1 DMA to the associated clock

This makes the kernel representation of the clock structure
match reality.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
So RP1's DMA block does have clock control, but it wasn't defined as a clock or used. That lead to the parent (`RP1_PLL_SYS_PRI_PH`) being disabled, which broke DMA.

Additional patches added to define rp1_clk_dma, and actually use it. I2S now works again.

P33M commented on drivers/clk/clk-rp1.c:
```diff
@@ -1874,6 +1869,29 @@ static const struct rp1_clk_desc clk_desc_array[] = {
+				.sel_reg = CLK_DMA_SEL,
+				.div_int_max = DIV_INT_8BIT_MAX,
+				.max_freq = 100 * MHz,
+				.fc0_src = FC_NUM(6, 7),
```
The index for this clock should be FC(2, 2) I believe.

6by9 replied:
I'll take your word for it as I have no information on this.

Updated.

P33M commented:
There are two remaining warts relating to SD/MMC - sdhci_core and sdio_src. There's "phy" in the sense that the PLL_SYS VCO frequency is directly used to generate sdio_src via rp1-clk-sdio, but sdhci_core is bogus.

However these interfaces are not used on Pi 5 so don't hold upstreaming.

6by9 commented:
Any further comments on this?

I do need to investigate why PLL_SYS_SEC needs to be critical, but we lose ethernet if it isn't, even though CLK_ETH is registering to use it. I haven't done a register comparison yet to figure out why, but that doesn't need to hold this up.

Fixing sdhci_core and sdio_src is a lower priority as the sdhci block isn't used (perhaps they should just be removed).

P33M commented:
Are there any other cases where a PLL output may go through an enabled -> Linux disables -> Linux attempts to enable cycle? Could be a bug in setup there, as that won't have been rigorously tested (PLLs are switched on by default).

P33M commented:
There's probably a bug here - 
```
CLK_IS_CRITICAL:
root@raspberrypi:/sys/kernel/debug/clk# cat pll_sys_pri_ph/regdump
ph_reg = 0x00051010
root@raspberrypi:/sys/kernel/debug/clk# cat pll_sys_sec/regdump
sec = 0x80000800
root@raspberrypi:/sys/kernel/debug/clk# cat pll_sys/regdump
prim = 0x00051010
root@raspberrypi:/sys/kernel/debug/clk# cat pll_sys_core/regdump
cs = 0x80000001
pwr = 0x00000004
fbdiv_int = 0x00000014
fbdiv_frac = 0x00000000

Clock isn't critical:
root@raspberrypi:/sys/kernel/debug/clk# cat pll_sys_pri_ph/regdump
ph_reg = 0x00051010
root@raspberrypi:/sys/kernel/debug/clk# cat pll_sys_sec/regdump
sec = 0x80000000
root@raspberrypi:/sys/kernel/debug/clk# cat pll_sys/regdump
prim = 0x00051010
root@raspberrypi:/sys/kernel/debug/clk# cat pll_sys_core/regdump
cs = 0x80000001
pwr = 0x00000004
fbdiv_int = 0x00000014
fbdiv_frac = 0x00000000
```
The secondary divider gets a bogus setup (0 = 19, if I remember).

P33M commented:
With no IS_CRITICAL flag, the divider gets turned on/off during boot. This tramples the divisor - and I never see a corresponding _set_rate call which is the only thing that could restore these bits. Surely if you're setting up a downstream clock, you demand that the parent produces the correct rate?

P33M commented:
This noddy patch that preserves the existing divisor keeps Ethernet working - 
```
diff --git a/drivers/clk/clk-rp1.c b/drivers/clk/clk-rp1.c
index 69ca9309b3b1..401356a236dd 100644
--- a/drivers/clk/clk-rp1.c
+++ b/drivers/clk/clk-rp1.c
@@ -927,7 +927,8 @@ static void rp1_pll_divider_off(struct clk_hw *hw)
        const struct rp1_pll_data *data = divider->data;

        spin_lock(&clockman->regs_lock);
-       clockman_write(clockman, data->ctrl_reg, PLL_SEC_RST);
+       clockman_write(clockman, data->ctrl_reg,
+                      clockman_read(clockman, data->ctrl_reg) | PLL_SEC_RST);
        spin_unlock(&clockman->regs_lock);
 }
```
but that glosses over the fact that PLL postdividers other than the ones downstream of pll_audio never get a set_rate call, so this just inherits the firmware clock setup.

6by9 commented:
No it doesn't solve the problem.

RP1_PLL_SYS_SEC was already in the list, so I was expecting that to have set the rate anyway, but it seems not. More research required on assigned-clocks at some point. #6479 to track that as it isn't critical.

Are all happy enough for now?

6by9 committed:
rp1: clk: Only set PLL_SEC_RST in rp1_pll_divider_off

Rather than clearing all the bits in rp1_pll_divider_off
and setting PLL_SEC_RST, retain the status of all the other
bits.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
RP1_PLL_SYS_SEC fixed up by not clearing register bits in rp1_pll_divider_off.

Branch rebased whilst at it.

That should close off all issues (I hope).
